subprojects/build-cache-packaging/src/test/groovy/org/gradle/caching/internal/packaging/impl/TarBuildCacheEntryPackerPermissionTest.groovy

```diff
@@ -35,29 +30,25 @@ class TarBuildCacheEntryPackerPermissionTest extends AbstractTarBuildCacheEntryP
 
     @Unroll
     def "can pack single file with file mode #mode"() {
-        def sourceOutputFile = Spy(File, constructorArgs: [temporaryFolder.file("source.txt").absolutePath]) as File
-        sourceOutputFile << "output"
-        def targetOutputFile = Spy(File, constructorArgs: [temporaryFolder.file("target.txt").absolutePath]) as File
-        def output = new ByteArrayOutputStream()
+        given:
+        def sourceOutputFile = temporaryFolder.file("source.txt") << "output"
         def unixMode = Integer.parseInt(mode, 8)
+        PermissionUtils.setPermissions(sourceOutputFile, unixMode)
+
+        and:
+        def targetOutputFile = temporaryFolder.file("target.txt")
+        def output = new ByteArrayOutputStream()
 
         when:
         pack output, prop(FILE, sourceOutputFile)
 
-        then:
-        1 * fileSystem.getUnixMode(sourceOutputFile) >> unixMode
-        _ * sourceOutputFile._
-        0 * _
-
-        when:
+        and:
         def input = new ByteArrayInputStream(output.toByteArray())
         unpack input, prop(FILE, targetOutputFile)
 
         then:
         targetOutputFile.text == "output"
-        1 * fileSystem.chmod(targetOutputFile, unixMode)
-        _ * targetOutputFile._
-        0 * _
+        PermissionUtils.permissions(targetOutputFile) == unixMode
```

Comment:
Will these tests pass on Windows as well?

Reply:
If you're asking like this I suspect not 😇 . Should we ignore it on Windows?

Reply:
Perhaps a better option would be to turn `PermissionUtils` into a service, so we could mock it here. Then we could add a separate test for Windows and Linux/macOS for this `PermissionSetter` service.

Reply:
Looking at the code again, it does something with `DosFilePermissions`. So maybe it even works on Windows...

Reply:
@lptr it passed the ready for merge pipeline so I guess it works on windows? If it wouldn't work on Windows I don't see much value in the additional abstraction to make something testable on Windows that does not work with the real implementation.

The advantage of having an additional interface would be that we could continue using `FileSystem` (via an adapter) instead of having the some functionality implemented in two places.

Reply:
I don't see any `Ready for Merge` builds executed for this branch, only the performance regression tests, and those don't involve Windows tests. I fired one off now: https://builds.gradle.org/viewQueued.html?itemId=30121516

If that build passes I'm okay with merging this as-is, and continue refining the API for `PermissionUtils` in a followup PR.

Reply:
Yep, fails on Windows: https://builds.gradle.org/viewLog.html?buildId=30121365

Reply:
We can do what you suggested, and put `FileSystem` behind a wrapper interface for Gradle, and use `PermissionUtils` in Maven.

Though that said I'd love to have a single implementation for the service using JDK 7 APIs, and some comprehensive tests to demonstrate what happens to permissions on Windows, Linux and macOS each. Maybe not something we need to do for this PR, though.

Reply:
Okay, so should we close this PR and create a new one introducing an interface?

I still don't get what you like to test on windows. As far as I understand the code does nothing on windows, no?

Reply:
I'm not 100% sure right now. Let's aim for keeping stuff as it is, and extract the interface, and use what we currently use underneath on both sides for now. That'll unblock us, and we can always come back to improve the situation if we think we need to.

Reply:
I will create a new PR for this.